subsys/net/lib/downloader/src/transports/http.c

```diff
@@ -0,0 +1,571 @@
+		}
+	} while (0);
+
+	p = strnstr(dl->config.buf, "\r\nconnection: close", parse_len);
+	if (p) {
+		LOG_WRN("Peer closed connection, will re-connect");
+		http->connection_close = true;
+	}
+
+	if (http->header.has_end) {
+		/* We have received the end of the header.
+		 * Verify that we have received everything that we need.
+		 */
+
+		if (!http->header.status_code) {
+			LOG_ERR("Server response malformed: status code not found");
+			return -EBADMSG;
+		}
+
+		expected_status = (http->ranged || dl->progress) ? 206 : 200;
+		if (http->header.status_code != expected_status) {
+			LOG_ERR("Unexpected HTTP response code %ld", http->header.status_code);
+			return -EBADMSG;
+		}
+
+		if (!dl->file_size) {
+			LOG_ERR("File size not set");
+			return -EBADMSG;
+		}
+
+		return parse_len;
+	}
+
+	q = dl->config.buf + buf_len;
+	/* We are still missing part of the header.
+	 * Return the lines (in number of bytes) that we have parsed.
+	 */
+	while (q > dl->config.buf && (*q != '\r') && (*q != '\n')) {
+		q--;
+	}
+
+	/* Keep \r and \n in the buffer in case it is part of the header ending. */
+	while (*(q - 1) == '\r' || *(q - 1) == '\n') {
+		q--;
+	}
+
+	parse_len = (q - dl->config.buf);
+
+	return parse_len;
+}
+
+/* Returns:
+ * Length of data payload left to process on success
+ * Negative errno on error.
+ */
+static int http_parse(struct downloader *dl, size_t len)
+{
+	int parsed_len;
+	struct transport_params_http *http;
+
+	http = (struct transport_params_http *)dl->transport_internal;
+
+	if (!http->header.has_end) {
+		/* Parse what we can from the header */
+		parsed_len = http_header_parse(dl, len);
+		if (parsed_len < 0) {
+			/* Something is wrong with the header */
+			return -EBADMSG;
+		}
+
+		if (parsed_len == len) {
+			dl->buf_offset = 0;
+			return 0;
+		} else if (parsed_len) {
+			/* Keep remaining payload */
+			len = len - parsed_len;
+			memmove(dl->config.buf, dl->config.buf + parsed_len, len);
+			dl->buf_offset = len;
+		}
+
+		if (!http->header.has_end) {
+			if (dl->config.buf_size == dl->buf_offset) {
+				LOG_ERR("Could not parse HTTP header lines from server (> %d)",
+					dl->config.buf_size);
+				return -E2BIG;
+			}
+			/* Wait for rest of header */
+			return 0;
+		}
+	}
+
+	/* Have we received a whole fragment or the whole file? */
+	if (dl->progress + len != dl->file_size) {
+		if (http->ranged) {
+			http->ranged_progress += len;
+			if (http->ranged_progress < (dl->host_config.range_override ?
+					   dl->host_config.range_override :
+					   TLS_RANGE_MAX - 1)) {
+				/* Ranged query: read until a full fragment */
+				return len;
+			}
+		} else {
+			/* Non-ranged query: just keep on reading, ignore fragment size */
+			return len;
+		}
+	}
+
+	/* Either we have a full file, or we need to request a next fragment */
+	http->new_data_req = true;
+	return len;
+}
+
+static bool dl_http_proto_supported(struct downloader *dl, const char *uri)
+{
+	if (strncmp(uri, "https://", 8) == 0) {
```

Comment:
These two strings and their length are used in multiple places, I would define them at the top and use `strlen`.